rust/lance-io/src/object_store/gcs_wrapper.rs

```diff
@@ -0,0 +1,458 @@
+                .await
+                .map_err(|source| UploadPutError {
+                    part_idx,
+                    buffer,
+                    source,
+                })?;
+            Ok((part_idx, part_id))
+        })
+    }
+
+    fn poll_tasks(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Result<(), io::Error> {
+        let mut_self = &mut *self;
+        loop {
+            match &mut mut_self.state {
+                UploadState::Pending | UploadState::Done => break,
+                UploadState::CreatingUpload(ref mut fut) => match fut.poll_unpin(cx) {
+                    Poll::Ready(Ok(multipart_id)) => {
+                        let futures = FuturesUnordered::new();
+                        let multipart_id = Arc::new(multipart_id);
+
+                        futures.push(Self::put_next_part(
+                            &mut_self.path,
+                            &mut_self.store,
+                            &mut mut_self.buffer,
+                            &mut 0,
+                            &multipart_id,
+                        ));
+
+                        mut_self.state = UploadState::InProgress {
+                            multipart_id,
+                            part_idx: 1, // We just used 0
+                            futures,
+                            part_ids: Vec::new(),
+                        };
+                    }
+                    Poll::Ready(Err(e)) => {
+                        return Err(std::io::Error::new(std::io::ErrorKind::Other, e))
+                    }
+                    Poll::Pending => break,
+                },
+                UploadState::InProgress {
+                    futures,
+                    part_ids,
+                    multipart_id,
+                    ..
+                } => {
+                    while let Poll::Ready(Some(res)) = futures.poll_next_unpin(cx) {
+                        match res {
+                            Ok((part_idx, part_id)) => {
+                                let total_parts = part_ids.len();
+                                part_ids.resize(total_parts.max(part_idx as usize + 1), None);
+                                part_ids[part_idx as usize] = Some(part_id);
+                            }
+                            Err(UploadPutError {
+                                source: OSError::Generic { source, .. },
+                                part_idx,
+                                buffer,
+                            }) if source.to_string().contains("Connection reset by peer")
+                                && mut_self.connection_resets < 20 =>
+                            {
+                                // Retry, but only up to 20 of them.
+                                mut_self.connection_resets += 1;
+
+                                // Resubmit with random jitter
+                                let sleep_time_ms = rand::thread_rng().gen_range(2_000..8_000);
```

Comment:
`2s` to `8s`?  That seems like quite a long retry interval

Reply:
True. The error we are dealing with here is mysterious but supposedly caused by some conditions on the network. I was thinking a long retry might be worthwhile. We might look at exponential backoff later to make this less conservative. I ran into this error a couple times while initially testing this, but after adding the retries with backoff, it worked smoothly.